plugin/pkg/admission/podpreset/admission.go

```diff
@@ -103,216 +104,217 @@ func (c *podPresetPlugin) Admit(a admission.Attributes) error {
 		return nil
 	}
 
-	list, err := c.lister.PodPresets(pod.GetNamespace()).List(labels.Everything())
-
 	// Ignore if exclusion annotation is present
 	if podAnnotations := pod.GetAnnotations(); podAnnotations != nil {
 		glog.V(5).Infof("Looking at pod annotations, found: %v", podAnnotations)
 		if podAnnotations[api.PodPresetOptOutAnnotationKey] == "true" {
 			return nil
 		}
 	}
+
+	list, err := c.lister.PodPresets(pod.GetNamespace()).List(labels.Everything())
 	if err != nil {
 		return fmt.Errorf("listing pod presets failed: %v", err)
 	}
 
-	// get the pod presets and iterate over them
-	for _, pip := range list {
-		selector, err := metav1.LabelSelectorAsSelector(&pip.Spec.Selector)
-		if err != nil {
-			return fmt.Errorf("listing pod presets for namespace:%s failed: %v", pod.GetNamespace(), err)
-		}
-
-		// check if the pod labels match the selector
-		if !selector.Matches(labels.Set(pod.Labels)) {
-			continue
-		}
-
-		glog.V(4).Infof("PodPreset %s matches pod %s labels", pip.GetName(), pod.GetName())
+	matchingPPs, err := filterPodPresets(list, pod)
+	if err != nil {
+		return fmt.Errorf("filtering pod presets failed: %v", err)
+	}
 
-		// merge in policy for Env
-		if pip.Spec.Env != nil {
-			for i, ctr := range pod.Spec.Containers {
-				r, err := mergeEnv(pip, ctr.Env)
-				if err != nil {
-					// add event to pod
-					c.addEvent(pod, pip, err.Error())
+	// detect merge conflict
+	err = safeToApplyPodPresetsOnPod(pod, matchingPPs)
+	if err != nil {
+		// conflict, ignore the error, but raise an event
+		glog.Warningf("conflict occurred while applying podpresets on pod: %v err: %v", pod.Name, err)
+		return nil
+	}
 
-					return nil
-				}
-				pod.Spec.Containers[i].Env = r
-			}
-		}
+	applyPodPresetsOnPod(pod, matchingPPs)
 
-		// merge in policy for EnvFrom
-		if pip.Spec.EnvFrom != nil {
-			for i, ctr := range pod.Spec.Containers {
-				r, err := mergeEnvFrom(pip, ctr.EnvFrom)
-				if err != nil {
-					// add event to pod
-					c.addEvent(pod, pip, err.Error())
+	glog.Infof("applied podpresets successfully on Pod: %s ", pod.GetGenerateName())
```

Comment:
Changed it to pod.Name. Also improved the logging a bit by making it more informative (added applied podpreset names as well)